src/memory_manager.cpp

```diff
@@ -102,20 +123,21 @@ void Manager::initialize(const rmmOptions_t* new_options) {
 
   if (nullptr != new_options) options = *new_options;
 
+  bool enable_logging = getOptions().enable_logging;
+
   if (usePoolAllocator()) {
     auto pool_size = getOptions().initial_pool_size;
     auto const& devices = getOptions().devices;
+
     if (useManagedMemory()) {
-      initialized_resource.reset(
-          new rmm::mr::cnmem_managed_memory_resource(pool_size, devices));
+      reset_resource(initialized_resource, new pool_managed_mr(pool_size, devices), enable_logging);
```

Comment:
When logging is enabled, the upstream resource won't get cleaned up when the `unique_ptr` is destroyed, it will only clean up the logger, and the logger doesn't do anything to cleanup the upstream. 

To solve this, we can either use a custom deleter in `reset_resource` that destroys both the upstream and the logger, or we can add another `unique_ptr<mr>` alongside `initialized_resource`. So something like this (but could use some streamlining):
```
   initialized_resource.reset(new cuda_mr{});
   if(enabled_logging){
      logging_adaptor.reset(new logging_resource_adaptor<cuda_memory_resource>(mr);
      set_default_resource(logging_adaptor.get());
   }
```

Reply:
It's funny I thought of this memory leak during your talk today too...

Reply:
Also just realized you already had a "do not merge" PR open for this!

Reply:
My solution was a hackier version of what you're doing. My PR _always_ enables logging whether the user requested it or not.

Reply:
Hey @jrhemstad I fixed this in my latest commit, but I'm not happy with it. I had to go out of my way to create the `reset_resource` template function just because I wanted to avoid a vtable lookup for every call to the logging_adaptor. This was made tricky because of the fact that I couldn't get memory_manager.hpp to compile if it includes `logging_resource_adaptor.hpp` due to the spdlog dependency (moreover, it's a template class, so it would be tricky to have a member pointer  to it anyway).

But the end result is I think the type is lost when it's stored into the unique_ptr member anyway, which means a user of get_default_resource will incur not one but two vtable lookups per call to allocate...

Reply:
Looking at the way you've implemented it, there will only be one virtual function call in `get_default_resource()->allocate(n)`. 

This will incur a vtable lookup to call `logging_resource_adaptor::allocate(n)`, but then when `logging_resource_adaptor::allocate(n)` calls `upstream.allocate(n)`, the upstream call will not incur a vtable lookup since the logging adaptor knows the concrete type of `upstream`. 

There may be a cleaner way to accomplish the same thing, but I think what you've got is fine given that it is intended to be temporary.

Reply:
Yes, good point. I was looking at the `logging_adaptor` member, since it is abstract.